File: encrypt_test.php
<?php

define('PW_MAGIC', 0xA3);
define('PW_FLAG',  0xFF);

function enc_byte($b) {
    $enc = (~($b ^ PW_MAGIC)) & 0xFF;
    return strtoupper(str_pad(dechex($enc), 2, '0', STR_PAD_LEFT));
}

function dec_next_char(&$s) {
    if (strlen($s) < 2) return 0;
    $base = "0123456789ABCDEF";
    $a = strpos($base, $s[0]);
    $b = strpos($base, $s[1]);
    $s = substr($s, 2);
    return (~(($a << 4) + $b) ^ PW_MAGIC) & 0xFF;
}

function encrypt($password, $key) {
    $full = $key . $password;
    $result = '';

    $result .= enc_byte(PW_FLAG);             // flag
    $result .= enc_byte(0);                   // dummy
    $result .= enc_byte(strlen($full));       // length
    $result .= enc_byte(0);                   // offset = 0

    for ($i = 0; $i < strlen($full); $i++) {
        $result .= enc_byte(ord($full[$i]));
    }

    return $result;
}

function decrypt($pwd, $key) {
    $clearpwd = "";

    $flag = dec_next_char($pwd);
    if ($flag == PW_FLAG) {
        dec_next_char($pwd); // dummy
        $length = dec_next_char($pwd);
    } else {
        $length = $flag;
    }

    $offset = dec_next_char($pwd);
    $pwd = substr($pwd, $offset * 2);

    for ($i = 0; $i < $length; $i++) {
        $clearpwd .= chr(dec_next_char($pwd));
    }

    if ($flag == PW_FLAG) {
        if (substr($clearpwd, 0, strlen($key)) !== $key) {
            return "";
        }
        $clearpwd = substr($clearpwd, strlen($key));
    }
    return $clearpwd;
}

function generate_random($opt) {
    $charset = '';
    if ($opt['useLowercase']) $charset .= 'abcdefghijklmnopqrstuvwxyz';
    if ($opt['useUppercase']) $charset .= 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    if ($opt['useDigits'])    $charset .= '0123456789';
    if ($opt['useSymbols'])   $charset .= '!@#$%^&*()-_=+[]{}|;:,.<>?';
    $charset .= $opt['extraChars'];

    $result = '';
    $max = strlen($charset) - 1;
    for ($i = 0; $i < $opt['length']; $i++) {
        $result .= $charset[random_int(0, $max)];
    }
    return $result;
}

function format_time($ts) {
    return date("Y-m-d H:i:s", $ts) . ':' . substr((string)(microtime(true) * 1000 % 1000 + 1000), 1, 3);
}

function format_duration($ms) {
    return floor($ms / 1000) . "s" . ($ms % 1000) . "ms";
}

// ====== Main Test ======
$user = "root";
$host = "192.168.12.34";
$key = $user . $host;

$opt = [
    'length' => 64,
    'useLowercase' => true,
    'useUppercase' => true,
    'useDigits' => true,
    'useSymbols' => true,
    'extraChars' => '#~-_'
];

$loopCount = 100;
$matchCount = 0;
$mismatchCount = 0;

$start = microtime(true);
echo "Start Time    : " . format_time(time()) . PHP_EOL;

for ($i = 1; $i <= $loopCount; $i++) {
    $password = generate_random($opt);
    $crypted = encrypt($password, $key);
    $plain = decrypt($crypted, $key);
    $match = ($password === $plain);
    if ($match) $matchCount++; else $mismatchCount++;

    echo "===== Test #$i =====\n";
    echo "Plain     : $password\n";
    echo "Encrypted : $crypted\n";
    echo "Decrypted : $plain\n";
    echo "Match     : " . ($match ? "......Yes" : "......No") . "\n\n";
    usleep(100000); // 100ms
}

$end = microtime(true);
$duration = intval(($end - $start) * 1000);
echo "End Time      : " . format_time(time()) . PHP_EOL;
echo "Total Duration: " . format_duration($duration) . PHP_EOL;
echo "Match Count   : $matchCount\n";
echo "Mismatch Count: $mismatchCount\n";

File: EncryptTest.java
import java.text.SimpleDateFormat;
import java.util.*;

public class EncryptTest {
    private static final int PW_MAGIC = 0xA3;
    private static final int PW_FLAG = 0xFF;

    public static String encrypt(String password, String key) {
        String full = key + password;
        StringBuilder result = new StringBuilder();

        result.append(encByte(PW_FLAG)); // enc_flag
        result.append(encByte(0));       // dummy
        result.append(encByte(full.length())); // length
        result.append(encByte(0));       // offset = 0

        for (char ch : full.toCharArray()) {
            result.append(encByte((int) ch));
        }

        return result.toString();
    }

    public static String decrypt(String hex, String key) {
        StringBuilder pwd = new StringBuilder(hex);
        StringBuilder clear = new StringBuilder();

        int flag = decNextChar(pwd);
        int length;

        if (flag == PW_FLAG) {
            decNextChar(pwd); // dummy
            length = decNextChar(pwd);
        } else {
            length = flag;
        }

        int offset = decNextChar(pwd);
        for (int i = 0; i < offset * 2 && i < pwd.length(); i++) {
            pwd.deleteCharAt(0);
        }

        for (int i = 0; i < length; i++) {
            clear.append((char) decNextChar(pwd));
        }

        if (flag == PW_FLAG) {
            if (!clear.toString().startsWith(key)) {
                return "";
            }
            return clear.substring(key.length());
        }
        return clear.toString();
    }

    private static String encByte(int b) {
        int enc = ~(b ^ PW_MAGIC) & 0xFF;
        return String.format("%02X", enc);
    }

    private static int decNextChar(StringBuilder s) {
        if (s.length() < 2) return 0;
        String base = "0123456789ABCDEF";
        int a = base.indexOf(s.charAt(0));
        int b = base.indexOf(s.charAt(1));
        int val = ~((a << 4) + b ^ PW_MAGIC) & 0xFF;
        s.delete(0, 2);
        return val;
    }

    public static String generateRandom(RandomOption opt) {
        StringBuilder charset = new StringBuilder();
        if (opt.useLowercase) charset.append("abcdefghijklmnopqrstuvwxyz");
        if (opt.useUppercase) charset.append("ABCDEFGHIJKLMNOPQRSTUVWXYZ");
        if (opt.useDigits) charset.append("0123456789");
        if (opt.useSymbols) charset.append("!@#$%^&*()-_=+[]{}|;:,.<>?");
        charset.append(opt.extraChars);

        Random rand = new Random();
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < opt.length; i++) {
            int idx = rand.nextInt(charset.length());
            result.append(charset.charAt(idx));
        }
        return result.toString();
    }

    public static String formatNow(Date date) {
        return new SimpleDateFormat("yyyy-MM-dd HH:mm:ss:SSS").format(date);
    }

    public static String formatDuration(long millis) {
        return String.format("%ds%dms", millis / 1000, millis % 1000);
    }

    public static void main(String[] args) throws InterruptedException {
        String user = "root";
        String host = "192.168.12.34";
        String key = user + host;

        RandomOption opt = new RandomOption();
        opt.length = 64;
        opt.useLowercase = true;
        opt.useUppercase = true;
        opt.useDigits = true;
        opt.useSymbols = true;
        opt.extraChars = "#~-_";

        int matchCount = 0, mismatchCount = 0;
        int loopCount = 100;

        long startMillis = System.currentTimeMillis();
        System.out.println("Start Time    : " + formatNow(new Date(startMillis)));

        for (int i = 1; i <= loopCount; i++) {
            String password = generateRandom(opt);
            String crypted = encrypt(password, key);
            String plain = decrypt(crypted, key);
            boolean match = password.equals(plain);
            if (match) matchCount++;
            else mismatchCount++;

            System.out.println("===== Test #" + i + " =====");
            System.out.println("Plain     : " + password);
            System.out.println("Encrypted : " + crypted);
            System.out.println("Decrypted : " + plain);
            System.out.println("Match     : " + (match ? "......Yes" : "......No"));
            System.out.println();

            Thread.sleep(100); // 等效 sleep 100ms
        }

        long endMillis = System.currentTimeMillis();
        System.out.println("End Time      : " + formatNow(new Date(endMillis)));
        System.out.println("Total Duration: " + formatDuration(endMillis - startMillis));
        System.out.println("Match Count   : " + matchCount);
        System.out.println("Mismatch Count: " + mismatchCount);
    }

    // 支持配置的随机生成结构体
    static class RandomOption {
        int length = 16;
        boolean useLowercase = true;
        boolean useUppercase = true;
        boolean useDigits = true;
        boolean useSymbols = false;
        String extraChars = "";
    }
}
